jupyterlab/handlers/ydoc.py

```diff
@@ -52,7 +61,25 @@ def source(self):
                             output["execution_count"] = int(execution_count)
         return dict(
             cells=cells,
-            metadata=meta["metadata"],
+            metadata=meta,
             nbformat=int(state["nbformat"]),
             nbformat_minor=int(state["nbformatMinor"]),
         )
+
+    @source.setter
+    def source(self, value):
+        with self._ydoc.begin_transaction() as t:
+            ycells = []
+            for cell in value["cells"]:
+                ycell = Y.YMap()
+                ycell.set(t, "source", cell["source"])
+                ycell.set(t, "metadata", cell["metadata"])
+                ycell.set(t, "cell_type", cell["cell_type"])
+                ycell.set(t, "id", cell["id"])
+                ycells.append(ycell)
+            self._ycells.push(t, ycells)
```

Comment:
We should probably delete all prior cells before pushing, even though it is only initialized once so it should not be a problem.

Reply:
I think it would be better to log an error and do nothing if this happens. Overwriting means basically dataloss. We expect that the document is empty.

Reply:
We also want to support changes that happen in the back-end, for instance when changing to another git branch changes the whole content of the file. We might want to ask the user what they want to do in that case (ignore the file change and save their changes, or loose their changes and revert to the saved file), but I think that we should support overwriting the Y model.